File: problems-adapt/0547_count-graph-components/solution_dfs.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {

    public int countComponents(int[][] adjacency) {
        int n = adjacency.length;
        boolean[] visited = new boolean[n];
        int components = 0;
        for (int start = 0; start < n; start++) {
            if (visited[start]) {
                continue;
            }
            // An unvisited city during the sweep starts a new component;
            // this one traversal absorbs exactly one component.
            components++;
            visited[start] = true;
            Deque<Integer> stack = new ArrayDeque<>();
            stack.push(start);
            while (!stack.isEmpty()) {
                int city = stack.pop();
                for (int other = 0; other < n; other++) {
                    if (adjacency[city][other] == 1 && !visited[other]) {
                        // Mark at push time so no city is stacked twice;
                        // membership is by visitation, so self-loops and the
                        // symmetric matrix never double count.
                        visited[other] = true;
                        stack.push(other);
                    }
                }
            }
        }
        return components;
    }
}

File: problems-adapt/0200-0300/0203_count-matrix-components/solution_dfs.py
class Solution:
    def countComponents(self, adjacency: list[list[int]]) -> int:
        n = len(adjacency)
        visited = [False] * n
        components = 0
        for start in range(n):
            if visited[start]:
                continue
            # An unvisited city during the sweep starts a new component;
            # this one traversal absorbs exactly one component.
            components += 1
            visited[start] = True
            stack = [start]
            while stack:
                city = stack.pop()
                for other in range(n):
                    if adjacency[city][other] == 1 and not visited[other]:
                        # Mark at push time so no city is stacked twice;
                        # membership is by visitation, so self-loops and the
                        # symmetric matrix never double count.
                        visited[other] = True
                        stack.append(other)
        return components
